### PR TITLE
Documentation overhaul

### LINKED ISSUE #421: Improve documentation
The documentation has grown over time and while very detailed in many places it is missing content in others and makes it difficult to find certain information. Another aspect that's been brought up multiple times is a better documentation of changes, see #140, #420 and #411.

This is an umbrella issue for everything related to restructuring and updating the documentation content. Below is a work in progress structure which is split into four sections; the Sketch app, the command line interface, plugin development and the file format.

Each section follows the same structure, introducing wider _Concepts_ and optionally providing _Guides_ for real world use cases which allows for a range of entry level to more advanced content.

### Sketch Developer
- Sketch app
  - Using the sketch:// URL scheme
- `sketchtool` CLI
  - Concepts
    - Versioning
    - Usage
  - Export
    - Pages
    - Artboards
    - Layers
    - Slices
    - Preview
  - Run plugins
  - List document structure
  - Show locations
  - Extract metadata
  - Dump file into JSON
- Plugin
  - Guides
    - Run Script within Sketch
    - Install a plugin
    - Remove a plugin
    - Create a Plugin
    - Debug a plugin
    - Publish a Plugin using skpm
    - Publish a Plugin manually
    - Update a Plugin
    - Handle an Action
    - Use an external text editor
    - Add a native user interface
    - Add a user interface using a webview
  - Concepts
    - Plugin Bundle
    - Sketch JavaScript Environment
    - Manifest
    - Sketch menu
    - Commands
    - Actions
    - API Versioning
  - JavaScript API
    - DOM
    - UI
    - Data
    - Settings and State Preservation
    - Actions
    - Changelog
  - Internal API
    - Disclaimer
    - CocoaScript
    - Sketch Headers
- File format
  - Concepts
    - Versioning
    - File Bundle
  - Sample file
- API Reference

### PR DESCRIPTION
**Pull request by christianklotz**
Closes #421 

- Reorganised structure 
- Removed outdated or incorrect content
- Added new guides

### CODE REVIEW

**Comment by bohemian-coding-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>This pull request may need a CHANGELOG entry.</td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  `Source/dom/models/Selection.js` changed, but not:
  - 📚 its docs
- 🧪 its tests

That's OK as long as you're refactoring.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: docs-website/pages/app.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+---
```
Comment (mathieudutour) [inline docs-website/pages/app.md]: I'd say it's a lot nicer to put the documentation in the `docs` folder and have the `docs-website` be the _code_ of the website, without actual content.

Comment (christianklotz) [reply docs-website/pages/app.md]: I see your point. I'll take another look at the file organisation once I've got the content in place

**Comment by mathieudutour** [comment]
I find the main page quite dry. There is no context whatsoever about what it is, what you can do, etc. For contrast, https://www.figma.com/developers eases you nicely into it.

File: docs-website/pages/plugins-debug-plugin.md [hunk 2]
```diff
@@ -1,11 +1,68 @@
...
+  - /introduction/preferences/
+
+order: 103
 ---
 
+When developing a Sketch plugin, chances are you will need some ways to know what is happening when your code is running.
+
+## Logs
+
+The most common way to debug a JavaScript code is to put a bunch of `console.log` at key steps. Unfortunately, JavaScriptCore (the [context in which a Sketch plugin is running](/guides/cocoascript/)) doesn't provide `console`. Instead, a special method called `log` is available.
```
Comment (mathieudutour) [inline docs-website/pages/plugins-debug-plugin.md]: That's not true anymore: `console.log` is polyfilled by Sketch

File: docs-website/pages/plugins-debug-plugin.md [hunk 3]
```diff
@@ -1,11 +1,68 @@
...
+
+mocha.protocols() // array of protocols the MSDocument class inherits from
+mocha.protocolsWithAncestors()
+```
+
+## Sketch-dev-tools
+
+We created a small Sketch specific tool to help you with debugging your plugins and hopefully make your life easier. It takes the form of a Sketch plugin that you can download [here](https://github.com/skpm/sketch-dev-tools/releases/latest) and launch with `cmd + option + j`.
+
 If you spend non-trivial amounts of time developing Plugins for Sketch, there are some improvements you can make to your workflow using these preferences.
```
Comment (mathieudutour) [inline docs-website/pages/plugins-debug-plugin.md]: we need a transition here

File: docs-website/pages/plugins-debug-plugin.md (hunk 2 above)
Comment (christianklotz) [reply docs-website/pages/plugins-debug-plugin.md]: I haven't touched any of the debugging content yet 🙂 Hopefully tomorrow

File: docs-website/pages/plugins-publish-plugin.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+       🚀 Select all shapes or just circles and rectangles.
+     </description>
+     <enclosure url="https://example.com/select-shapes.sketchplugin-1.0.zip" sparkle:version="1.0"/>
+   </item>
+   </channel>
+   </rss>
+   ```
+1. To include the plugin in the official listing, open a pull request on [`plugin-directory`](https://github.com/sketchplugins/plugin-directory) providing information about your plugin including the URL to the appcast.
+
+> **Note:** To use rich text in the release notes, include HTML formatted `description`, e.g. `<![CDATA[<h1>New 🎉</h1>]]>`.
```
Comment (mathieudutour) [inline docs-website/pages/plugins-publish-plugin.md]: that's not supported yet, Sketch doesn't show any release notes for plugins

File: docs-website/pages/plugins-update-plugin.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+1. Create a GitHub release.
+1. Update the `.appcast.xml` with the latest release information.
+
+Run the following command providing either a specific version number of bump the existing version by `path`, `minor` or `major`. For more available options see `skpm publish --help`.
+
+```sh
+skpm publish <version>
+```
+
+> **Note:** If the previous version was not included in the official listing before `skpm` will also open a pull request on [`plugin-directory`](https://github.com/sketchplugins/plugin-directory) to submit it unless the command is run with the `--skip-registry` command-line argument.
```
Comment (mathieudutour) [inline docs-website/pages/plugins-update-plugin.md]: or if there is `"private": true` in the package.json

File: docs-website/pages/plugins-debug-plugin.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+```js
+let mocha = context.document.class().mocha()
+console.log(mocha.properties())
+```
+
+See the [Mocha documentation](https://github.com/logancollins/Mocha#introspecting-the-objective-c-runtime) for a complete list of available introspection methods.
+
+## Inspect Sketch documents with Sketch DevTools
+
+[Sketch DevTools](https://github.com/skpm/sketch-dev-tools) is a Sketch plugin that lets you inspect elements within a document without writing code.
```
Comment (mathieudutour) [inline docs-website/pages/plugins-debug-plugin.md]: It also allow to see logs in a more convenient way (eg. you can fold/unfold arrays and objects, etc.).
It also introspect the obj-c objects directly, so if you log `context.document`, you can see its properties and methods without `class().mocha()`.
It also shows all the actions that are happening

File: docs-website/pages/plugins-debug-plugin.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+```sh
+defaults write com.bohemiancoding.sketch3.beta WebKitDeveloperExtras -bool YES
+```
+
+DOM elements can be inspected by right-clicking inside the webview and select _Inspect Element_ which brings up the Safari Web Inspector. If you have JavaScript code supressing the context menu you can still inspect elements by selecting them directly within _Elements_ in Safari Web Inspector.
+
+> **Note:** If you're using `skpm`, the `WebKitDeveloperExtras` user default will be set to `YES` automatically.
+
+## Troubleshooting
```
Comment (mathieudutour) [inline docs-website/pages/plugins-debug-plugin.md]: It feels weird to have this section in "Debugging a plugin", there are a bit of everything without much link to the previous section?

File: docs-website/pages/plugins.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+~/Library/Application Support/com.bohemiancoding.sketch3/PluginsWarehouse
+```
+
+> **Quick tip:** You can access the plugins folder by holding `alt` while opening the _Plugins_ menu in Sketch and then choose 'Reveal Plugins Folder'.
+
+### Install a plugin
+
+See the [Sketch documentation](https://sketch.com/docs/plugins/) on how to install plugins.
+
+For development
```
Comment (KevinGutowski) [inline docs-website/pages/plugins.md]: Seems like this line was cut off?

File: docs-website/pages/plugins.md [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+To remove a Plugin, just select the **Plugins › Manage Plugins…** menu option, select the Plugin you want to remove from the list, and either right click the plugin or click the gear icon and select _Uninstall "Plugin Name"_:
+
+![](/images/developer/plugin-uninstall.png)
+
+Any commands that the Plugin provided will be removed from the **Plugins** menu immediately.
+
+Alternatively, you can just uncheck any Plugin on the list to disable it without uninstalling it.
+
+## JavaScript API
+
```
Comment (KevinGutowski) [inline docs-website/pages/plugins.md]: Seems like this is still left as a TODO but just want to be sure.

File: docs-website/pages/plugins-publish-plugin.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+Run `skpm` from your plugin folder.
+
+```sh
+skpm publish
+```
+
+For all options such as not including a plugin in the official listing view the usage instructions.
+
+```sh
+skpm publish help
```
Comment (KevinGutowski) [inline docs-website/pages/plugins-publish-plugin.md]: I’m pretty sure the command is `skpm publish --help`

File: docs-website/pages/plugins-manifest.md [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+#### `description`
+
+Provides a general description of what the plugin does.
+
+#### `disableCocoaScriptPreprocessor`
+
+Defines if the CocoaScript preprocessor should be used for the plugin's scripts. Set this to `true` to use ES6 syntax and build systems like [skpm](https://skpm.io).
+
+When disabled, `@import` and bracket syntax, i.e. `[obj hello: world]` are no longer supported.
```
Comment (KevinGutowski) [inline docs-website/pages/plugins-manifest.md]: this line is a little confusing because of the parameter name `disableCocoaScriptPreprocessor`

Maybe, 

"When set to `true`, `@import` and bracket syntax, i.e. `[obj hello: world]` are supported"

File: docs-website/pages/plugins-js-env.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+- [`process`](https://github.com/skpm/process)
+- [`querystring`](https://github.com/skpm/querystring)
+- [`stream`](https://github.com/skpm/stream)
+- [`string_decoder`](https://github.com/skpm/string_decoder)
+- [`timers`](https://github.com/skpm/timers)
+- [`util`](https://github.com/skpm/util)
+
+You can find all the official polyfills [on Github](https://github.com/search?q=topic%3Apolyfill+org%3Askpm&type=Repositories).
+
+> To see which modules are installed for a specific Sketch version, head to [`core-modules/package.json`](https://github.com/BohemianCoding/SketchAPI/blob/develop/core-modules/package.json) within the Sketch API and select the release branch for the version e.g. `release/53.2`.
```
Comment (KevinGutowski) [inline docs-website/pages/plugins-js-env.md]: Perhaps it would be helpful to note that the first polyfills were initially added with Sketch 50.

File: docs-website/pages/plugins-bundle.md [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+```
+
+| Folder      | Description                                                                              |
+| ----------- | ---------------------------------------------------------------------------------------- |
+| `Contents`  | Contains the manifest and all JavaScript (`.js`) and CocoaScript (`.cocoascript`) files. |
+| `Resources` | Contains all of the plugin's resource files.                                             |
+
+## Plugin Manifest file
+
+For Sketch to initialize the plugin correctly, the `manifest.json` must be included in the plugin bundle. It defines how the scripts can be called from either the menu or a keyboard shortcut, or triggered by user activity within Sketch. For more information, see the [manifest documentation](/plugins/manifest).
```
Comment (KevinGutowski) [inline docs-website/pages/plugins-bundle.md]: Broken link. Should be `[manifest documentation](/plugins/plugin-manifest)`

File: docs-website/pages/plugins-js.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+In the manifest file, you specify a dictionary describing each command that the Plugin defines.
+
+In this dictionary, _script_ and _handler_ keys tell Sketch which script file to look in, and which handler to run.
+
+You are free to put each command implementation into its own script file, or to put them all in a single file.
+
+You must specify the _script_ key for each command.
+
+If you put each command in its own script file, you can omit the _handler_ key. In this case, Sketch will default to calling the `export default function () {}` handler.
```
Comment (KevinGutowski) [inline docs-website/pages/plugins-js.md]: There is a lot of text here that makes it a little hard to follow/visualize. Perhaps provide an example structure here or maybe the simplest thing to do is to link back to the manifest page.

File: docs-website/pages/plugins-actions.md [hunk 15]
```diff
@@ -1,8 +1,14 @@
...
+chapter: JavaScript API
+permalink: /plugins/actions
+
+redirect_from:
+  - /guides/action-api/
+
+order: 304
 ---
 
 In Sketch 3.8, we introduced the Action API: a way to let plugins react to events in the app. Using it, plugin authors can write code that is executed when certain actions are triggered, like "opening a document", "saving", "adding an artboard"…
```
Comment (KevinGutowski) [inline docs-website/pages/plugins-actions.md]: On this page it might make sense to communicate that the dev tools provide a really easy way to listen for all actions.

Screen Shot 2019-04-20 at 12 20 29 AM

**Comment by KevinGutowski** [comment]
Screen Shot 2019-04-20 at 12 27 36 AM

Is it intentional for the Actions Page to have the sidebar filled with all the links? Seems like on the deployed branch it is different.
Screen Shot 2019-04-20 at 12 36 58 AM

File: docs-website/pages/cli-run-plugin.md [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+context.document.showMessage("Remote plugin running!")
+```
+
+We save it as 'Remote Plugin' from the _Run Script…_ panel, and run this:
+
+```
+sketchtool run ~/Library/Application\ Support/com.bohemiancoding.sketch3/Plugins/Remote\ Plugin.sketchplugin com.bohemiancoding.sketch.runscriptidentifier
+```
+
+Note that `com.bohemiancoding.sketch.runscriptidentifier` is the default command identifier Sketch uses when saving a plugin, but it may be different in your case. If you just want to run the first command in the bundle, you can use `""` instead of an identifier.
```
Comment (KevinGutowski) [inline docs-website/pages/cli-run-plugin.md]: I’m a little unclear how to run the first command in the bundle.

is it `com.bohemiancoding.sketch.""`

File: docs-website/pages/plugins-publish-plugin.md (hunk 4 above)
Comment (christianklotz) [reply docs-website/pages/plugins-publish-plugin.md]: Oh, ok … thank you for flagging

**Comment by bohemian-coding-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  `Source/dom/Factory.js` changed, but not:
  - 🧪 its tests

That's OK as long as you're refactoring.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  `Source/dom/layers/Group.js` changed, but not:
  - 📚 its docs
- 🧪 its tests

That's OK as long as you're refactoring.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  `Source/dom/layers/SymbolMaster.js` changed, but not:
  - 🧪 its tests

That's OK as long as you're refactoring.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  `Source/dom/models/CurvePoint.js` changed, but not:
  - 📚 its docs

That's OK as long as you're refactoring.
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  `Source/ui/UI.js` changed, but not:
  - 🧪 its tests

That's OK as long as you're refactoring.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 7cca6de26cc5621223cb2f7a2a2131030ae5fe09
</p>

**Comment by christianklotz** [comment]
Closing for #498 that tidies up the git history
